contract-message-test/tests/utils/ext_sdk_provider.rs

```diff
@@ -0,0 +1,164 @@
+/**
+ * TODO: This module contains functions that should eventually
+ * be made part of the fuels-rs sdk repo as part of the Provider
+ * implementation, similar to functions like 'build_transfer_tx'
+ */
+use fuel_crypto::Hasher;
+
+use fuels::prelude::*;
+use fuels::tx::{Address, AssetId, Bytes32, Contract as tx_contract, Input, Output, Transaction};
+
+const CONTRACT_MESSAGE_MIN_GAS: u64 = 1_200_000;
+const CONTRACT_MESSAGE_SCRIPT_BINARY: &str =
+    "../contract-message-script/out/debug/contract_message_script.bin";
+const CONTRACT_MESSAGE_PREDICATE_BINARY: &str =
+    "../contract-message-predicate/out/debug/contract_message_predicate.bin";
+
+/// Gets the message to contract script
+pub async fn get_contract_message_script() -> (Vec<u8>, Bytes32) {
+    let script_bytecode = std::fs::read(CONTRACT_MESSAGE_SCRIPT_BINARY).unwrap();
+    let script_hash = Hasher::hash(&script_bytecode.clone());
+    (script_bytecode, script_hash)
+}
+
+/// Gets the message to contract predicate
+pub async fn get_contract_message_predicate() -> (Vec<u8>, Address) {
+    let predicate_bytecode = std::fs::read(CONTRACT_MESSAGE_PREDICATE_BINARY).unwrap();
+    let predicate_root = Address::from(*tx_contract::root_from_code(&predicate_bytecode));
+    (predicate_bytecode, predicate_root)
+}
+
+/// Build a message-to-contract transaction with the given input coins and outputs
+/// note: unspent gas is returned to the owner of the first given gas input
+pub async fn build_contract_message_tx(
+    message: Input,
+    contract: Input,
+    gas_coins: &[Input],
+    optional_inputs: &[Input],
+    optional_outputs: &[Output],
+    params: TxParameters,
+) -> Transaction {
+    // Get the script and predicate for contract messages
+    let (script_bytecode, _) = get_contract_message_script().await;
+
+    // Start building tx list of inputs
+    let mut tx_inputs: Vec<Input> = Vec::new();
+    tx_inputs.push(contract);
+    tx_inputs.push(message);
+
+    // Start building tx list of outputs
+    let mut tx_outputs: Vec<Output> = Vec::new();
+    tx_outputs.push(Output::Contract {
+        input_index: 0u8,
+        balance_root: Bytes32::zeroed(),
+        state_root: Bytes32::zeroed(),
+    });
+
+    // Build a change output for the owner of the first provided coin input
+    if !gas_coins.is_empty() {
+        let coin: &Input = &gas_coins[0];
+        match coin {
```

Comment:
I'm not sure how to do this. It's checking if `Input::CoinSigned` **_or_** `Input::CoinPredicate`. I'm not sure if that's even possible with an `if let`. If there is, let me know!

Reply:
I still can't get this to compile using just an `if`. I think it's complicated because the enum is using structs and I need to get the `owner` field from the enum as well as part of the logic. Everything I'm seeing online says to use `match`

Reply:
I would have to double-check, but I don't think this type of comparison can be done with an `if` ?
I thought only `match` or `if let` work for enums... 🤔

Reply:
So, `if` could be made to work here, but only _if_ you `impl Eq for Input {...`

Reply:
How would you get the `owner` field in either `CoinInput` or `CoinPredicate` inside the if statement? Isn't it easier to just use match?